File: poly_regression.py
# -*- coding: utf-8 -*-
"""
Created on Tue Sep 19 08:54:25 2017

@author: dhanu
"""
#import required libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
#importing the dataset
dataset=pd.read_csv('Position_Salaries.csv')
X=dataset.iloc[:,1:2].values
Y=dataset.iloc[:,2].values
"""#training and testing the dataset
from sklearn.cross_validation import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(X, Y, test_size=0.2, random_state=0)"""

#Fitting Linear Regression to the dataset
from sklearn.linear_model import LinearRegression
lin_reg=LinearRegression()
lin_reg.fit(X,Y)
#Fitting Polyniómial regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg=PolynomialFeatures(degree=4)
X_poly=poly_reg.fit_transform(X)
lin_reg2=LinearRegression()
lin_reg2.fit(X_poly,Y)
#Visualizing the Linear Regression model
plt.scatter(X,Y, color='red')
plt.plot(X,lin_reg.predict(X), color='green')
plt.title('Truth or Bluff(Linear Regression)')
plt.xlabel('position level')
plt.ylabel('Salary')
plt.show()
#Visualize the polynomial regression model
X_grid = np.arange(min(X),max(X),0.1)
X_grid=X_grid.reshape(len(X_grid),1)
plt.scatter(X,Y, color='red')
plt.plot(X_grid,lin_reg2.predict(poly_reg.fit_transform(X_grid)), color='green')
plt.title('Truth or Bluff(Polynomial Regression)')
plt.xlabel('position level')
plt.ylabel('Salary')
plt.show()
#Increase the degree by 1 to get the exact curve and the exact results.
#predict a new result with linear regression model
lin_reg.predict(6.5)
#predict a new value with polynomial regression model
lin_reg2.predict(poly_reg.fit_transform(6.5))

File: poly_regression.R
#import dataset
dataset = read.csv('Position_Salaries.csv')
# dataset = dataset[, 2:3]
#training and testing sets
#install.packages('caTools')
# library(caTools)
# set.seed(123)
# split=sample.split(dataset$Purchased,SplitRatio = 0.8)
# training_set=subset(dataset,split==TRUE)
# testing_set=subset(dataset,split==FALSE)
#feature scaling
# training_set[,2:3]=scale(training_set[,2:3])
# testing_set[,2:3]=scale(testing_set[,2:3])
#Fitting Linear Regression to the dataset
lin_reg=lm(formula=Salary~.,
           data=dataset)
#Fitting Polynomial Regression to the dataset
dataset$Level2=dataset$Level^2
dataset$Level3=dataset$Level^3
dataset$Level4=dataset$Level^4
poly_reg=lm(formula=Salary~.,
            data=dataset)
#Visualizing the LinearRegression results
library(ggplot2)
ggplot()+
  geom_point(aes(x=dataset$Level,y=dataset$Salary),
             colour='red')+
  geom_line(aes(x=dataset$Level,y=predict(lin_reg, newdata = dataset)),
            colour='blue')+
  ggtitle('Truth or Bluff(Linear Regression )')+
  xlab('Level')+
  ylab('Salary')
#visualizing the Polynomial Regression results
ggplot()+
  geom_point(aes(x=dataset$Level,y=dataset$Salary),
             colour='red')+
  geom_line(aes(x=dataset$Level,y=predict(poly_reg, newdata = dataset)),
            colour='blue')+
  ggtitle('Truth or Bluff(Linear Regression )')+
  xlab('Level')+
  ylab('Salary')
#Predicitng a new result with Linear Regression
y_pred=predict(lin_reg,data.frame(Level=6.5))
#Predicting a new result with Polynomial Regression
y_pred=predict(poly_reg,data.frame(Level=6.5,
                                   Level2=6.5^2,
                                   Level3=6.5^3,
                                   Level4=6.5^4))
